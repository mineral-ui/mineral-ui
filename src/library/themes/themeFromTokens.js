/* @flow */
import pxToEm from '../styles/pxToEm';
import colorAliases from './generated/colorAliases';

type Aliases = { [string]: string };
type ColorOptions = {
  aliases?: Aliases,
  color: Ramp | string,
  colorName: string,
  newKey?: (string) => string,
  tokens: ColorTokens
};
type ColorTheme = ColorTokens;
type ColorTokens = { [string]: string };
export type Ramp = { [number]: string };
type Theme = Tokens;
type Tokens = { [string]: number | string };

const contains = (string: string, subString: string) =>
  string.indexOf(subString) !== -1;

const newKey = (key) => key.replace('brand', 'theme');

const remToEm = (value: string) => value.replace('rem', 'em');

export default function(tokens: Tokens): Theme {
  return Object.keys(tokens).reduce((acc, key) => {
    let value = tokens[key];
    if (typeof value === 'string') {
      if (value.split('px').length === 2 && !contains(key, 'breakpoint')) {
        value = pxToEm(value);
      } else if (contains(key, 'fontSize')) {
        value = remToEm(value);
      }
    }

    acc[newKey(key)] = value;
    return acc;
  }, {});
}

export function themeFromColorTokens({
  aliases = colorAliases,
  color,
  colorName,
  tokens
}: ColorOptions): ColorTheme {
  return Object.keys(tokens).reduce((acc, key) => {
    let value = tokens[key];
    if (typeof color === 'string') {
      value = color;
    } else {
      const rampKey = parseInt(aliases[key].split('_')[1]);
      if (rampKey) {
        if (color[rampKey]) {
          value = color[rampKey];
        } else {
          throw new Error(`[mineral-ui/themes/createTheme]: colors.${colorName}[${rampKey}] is missing.
  See https://github.com/mineral-ui/mineral-ui/blob/master/packages/mineral-ui-tokens/src/blue.js for an example color ramp.`);
        }
      }
    }

    acc[newKey(key)] = value;
    return acc;
  }, {});
}
