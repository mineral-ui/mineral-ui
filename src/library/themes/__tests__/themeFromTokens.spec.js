/* @flow */
import themeFromTokens, { themeFromColorTokens } from '../themeFromTokens';

const ALIASES = {
  color_black: 'black',
  color_brand: 'brand_10',
  color_danger: 'red_20',
  color_gray: 'gray_30',
  color_white: 'white'
};
const BAD_RAMP = {
  [1]: 'bad'
};
const CUSTOM_RAMP = {
  [10]: 'custom_brand',
  [20]: 'custom_danger',
  [30]: 'custom_gray',
  [40]: '40',
  [50]: '50',
  [60]: '60',
  [70]: '70',
  [80]: '80',
  [90]: '90',
  [100]: '100'
};
const TOKENS_BLACK = {
  color_black: 'black'
};
const TOKENS_BRAND = {
  color_brand: 'blue'
};
const TOKENS_DANGER = {
  color_danger: 'red'
};
const TOKENS_DEFAULT = {
  color_theme: 'blue',
  pixel: '16px',
  boxShadow: '1px 1px 1px rgba(0,0,0,0.1)',
  breakpoint: '16px',
  fontSize_test: '1rem',
  untouched: 100
};
const TOKENS_GRAY = {
  color_gray: 'gray'
};
const TOKENS_WHITE = {
  color_white: 'white'
};

const themeFromColorTokensWithAliases = (color, tokens) =>
  themeFromColorTokens({
    aliases: ALIASES,
    color,
    colorName: 'unimportant',
    tokens
  });

describe('themeFromTokens', () => {
  describe('converts', () => {
    const theme = themeFromTokens({ tokens: TOKENS_DEFAULT });
    it('rems to ems', () => {
      expect(theme.fontSize_test).toMatchSnapshot();
    });

    it(`'brand' to 'theme' in the key`, () => {
      expect(theme.color_theme).toMatchSnapshot();
      expect(theme.color_brand).toBeUndefined();
    });

    it('leaves some stuff alone', () => {
      expect(theme.boxShadow).toMatchSnapshot();
      expect(theme.untouched).toMatchSnapshot();
    });

    describe('px to ems', () => {
      it('when not a breakpoint', () => {
        expect(theme.pixel).toMatchSnapshot();
      });

      it(`except when key contains 'breakpoint'`, () => {
        expect(theme.breakpoint).toMatchSnapshot();
      });
    });
  });
});

describe('themeFromTokens', () => {
  describe('with colors', () => {
    describe('with black', () => {
      it('overrides black value', () => {
        const theme = themeFromColorTokensWithAliases(
          'custom_black',
          TOKENS_BLACK
        );
        expect(theme.color_black).toMatchSnapshot();
      });
    });

    describe('with intent', () => {
      it('overrides intent value', () => {
        const theme = themeFromColorTokensWithAliases(
          CUSTOM_RAMP,
          TOKENS_DANGER
        );
        expect(theme.color_danger).toMatchSnapshot();
      });
    });

    describe('with gray', () => {
      it('overrides gray value', () => {
        const theme = themeFromColorTokensWithAliases(CUSTOM_RAMP, TOKENS_GRAY);
        expect(theme.color_gray).toMatchSnapshot();
      });
    });

    describe('with theme', () => {
      it('overrides theme value', () => {
        const theme = themeFromColorTokensWithAliases(
          CUSTOM_RAMP,
          TOKENS_BRAND
        );
        expect(theme.color_theme).toMatchSnapshot();
      });
    });

    describe('with white', () => {
      it('overrides white value', () => {
        const theme = themeFromColorTokensWithAliases(
          'custom_white',
          TOKENS_WHITE
        );
        expect(theme.color_white).toMatchSnapshot();
      });
    });

    describe('with invalid ramp as any color', () => {
      it('throws error', () => {
        expect(() => {
          themeFromColorTokensWithAliases(BAD_RAMP, TOKENS_DANGER);
        }).toThrowError();
      });
    });
  });
});
