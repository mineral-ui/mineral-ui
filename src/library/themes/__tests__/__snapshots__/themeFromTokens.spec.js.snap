// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`themeFromTokens converts 'brand' to 'theme' in the key 1`] = `"blue"`;

exports[`themeFromTokens converts leaves some stuff alone 1`] = `"1px 1px 1px rgba(0,0,0,0.1)"`;

exports[`themeFromTokens converts leaves some stuff alone 2`] = `100`;

exports[`themeFromTokens converts px to ems except when key contains 'breakpoint' 1`] = `"16px"`;

exports[`themeFromTokens converts px to ems when not a breakpoint 1`] = `"1em"`;

exports[`themeFromTokens converts rems to ems 1`] = `"1em"`;

exports[`themeFromTokens with colors with black overrides black value 1`] = `"custom_black"`;

exports[`themeFromTokens with colors with gray overrides gray value 1`] = `"custom_gray"`;

exports[`themeFromTokens with colors with intent overrides intent value 1`] = `"custom_danger"`;

exports[`themeFromTokens with colors with theme overrides theme value 1`] = `"custom_brand"`;

exports[`themeFromTokens with colors with white overrides white value 1`] = `"custom_white"`;
